### PR TITLE
seccomp: implement mount syscall interception

### PR DESCRIPTION
**Pull request by brauner**
This allows to intercept and subsequently to emulate the mount syscall.
To implement an efficient seccomp filter, we filter based on the flags
argument of the mount syscall. We don't want to filter any of the
following flag combinations since they do not cause the creation of a
new superblock:

```
MS_REMOUNT
MS_BIND
MS_MOVE
MS_UNBINDABLE
MS_PRIVATE
MS_SLAVE
MS_SHARED
MS_KERNMOUNT
MS_I_VERSION

So define the following mask of allowed flags:

long unsigned int mask = MS_MGC_VAL | MS_RDONLY | MS_NOSUID | MS_NODEV |
                         MS_NOEXEC | MS_SYNCHRONOUS | MS_MANDLOCK |
                         MS_DIRSYNC | MS_NOATIME | MS_NODIRATIME | MS_REC |
                         MS_VERBOSE | MS_SILENT | MS_POSIXACL | MS_RELATIME |
                         MS_STRICTATIME | MS_LAZYTIME;

Now we inverse the flag:

inverse_mask ~= mask;
```

Seccomp will now only intercept these flags if they do not contain any of
the allowed flags, i.e. we only intercept combinations were a new superblock
is created.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
As discussed this morning, I believe we want three config options:
 - security.syscalls.intercept.mount (boolean)
 - security.syscalls.intercept.mount.allowed (list of comma separated strings)
 - security.syscalls.intercept.mount.fuse (list of comma separate strings)

A filesystem cannot be listed in both `allowed` and `fuse`, config validation should prevent it.

The syntax for `allowed` is a comma separate list of filesystem names, any filesystem in that list will be mounted through interception when requested.

The syntax for `fuse` is a comma separated list of filesystem names and fuse binaries, like "squashfs=squashfuse". The first argument passed to the binary will be the mount source, the second will be the mount destination. For the time being, we will discard all options passed to the mount syscall as there isn't a universal way to handle this in fuse.

@brauner for the shiftfs integration, what do you think we should do? Add some modifier to `allowed`, something like `ext4+shiftfs` to indicate that `ext4` is getting shiftfs on top of it, or have a global `security.syscalls.intercept.mount.shift` that we set to `true` to indicate that all `allowed` mounts should have shiftfs applied to them?

**Comment by brauner** [comment]
> As discussed this morning, I believe we want three config options:
> 
> * security.syscalls.intercept.mount (boolean)
> * security.syscalls.intercept.mount.allowed (list of comma separated strings)
> * security.syscalls.intercept.mount.fuse (list of comma separate strings)
> 
> A filesystem cannot be listed in both `allowed` and `fuse`, config validation should prevent it.
> 
> The syntax for `allowed` is a comma separate list of filesystem names, any filesystem in that list will be mounted through interception when requested.
> 
> The syntax for `fuse` is a comma separated list of filesystem names and fuse binaries, like "squashfs=squashfuse". The first argument passed to the binary will be the mount source, the second will be the mount destination. For the time being, we will discard all options passed to the mount syscall as there isn't a universal way to handle this in fuse.
> 
> @brauner for the shiftfs integration, what do you think we should do? Add some modifier to `allowed`, something like `ext4+shiftfs` to indicate that `ext4` is getting shiftfs on top of it, or have a global `security.syscalls.intercept.mount.shift` that we set to `true` to indicate that all `allowed` mounts should have shiftfs applied to them?

I'd vote for the latter option.

**Comment by stgraber** [comment]
Ok, lets go with that then, can you implement that bit too?

We'll want to effectively mount+mark the filesystem somewhere, then mount shiftfs on top, then unmount the underlay. That leaves us with a single mount entry in the container which can be unmounted normally by the user (that's what we do for mount injection).

**Comment by brauner** [comment]
I've implemented all bits apart from `security.syscalls.intercept.mount.fuse`. I want to delay this because it makes things way more complex and I want time to think this through. This shouldn't block the core patchset though.

**Comment by stgraber** [comment]
@brauner what's the difficulty around fuse?

My understanding is that in this case we want to perfectly match the caller task (all namespaces, all caps, cwd, root, ...), double fork, exec the binary and if that all went well, return 0 back to caller.

**Comment by brauner** [comment]
> Branch is looking a bit red :)

Good, my cgo compilation fixes are catching unused variables and other things. :)

**Comment by brauner** [comment]
> @brauner what's the difficulty around fuse?
> 
> My understanding is that in this case we want to perfectly match the caller task (all namespaces, all caps, cwd, root, ...), double fork, exec the binary and if that all went well, return 0 back to caller.

I'm wondering whether we can just get away with using `mount.fuse` instead of individual binaries and like to experiment a bit.

**Comment by stgraber** [comment]
Needs rebase

**Comment by brauner** [comment]
> Needs rebase

Rebased.

File: doc/api-extensions.md [hunk 1]
```diff
@@ -859,3 +859,10 @@ This adds support for an optional argument (`ceph.osd.data_pool_name`) when crea
 storage pools using Ceph RBD, when this argument is used the pool will store it's
 actual data in the pool specified with `data_pool_name` while keeping the metadata
 in the pool specified by `pool_name`.
+
+## container\_syscall\_intercept\_mount
```
Comment (stgraber) [inline doc/api-extensions.md]: Needs to be in a `api: Add container_syscall_intercept_mount extension` commit

File: doc/containers.md [hunk 2]
```diff
@@ -34,57 +34,60 @@ currently supported:
...
+security.syscalls.blacklist\_default            | boolean   | true              | no            | container\_syscall\_filtering        | Enables the default syscall blacklist
+security.syscalls.intercept.mknod               | boolean   | false             | no            | container\_syscall\_intercept        | Handles the `mknod` and `mknodat` system calls (allows creation of a limited subset of char/block devices)
+security.syscalls.intercept.mount               | boolean   | false             | no            | container\_syscall\_intercept\_mount | Handles the `mount` system call
+security.syscalls.intercept.mount.allowed       | string    | -                 | yes           | container\_syscall\_intercept\_mount | Specify a comma-separated list of filesystems that are safe to mount for processes inside the container.
+security.syscalls.intercept.mount.shift         | boolean   | false             | yes           | container\_syscall\_intercept\_mount | Whether to mount shiftfs on top of filesystems handled through mount syscall interception.
```
Comment (stgraber) [inline doc/containers.md]: This change should be in a `doc: Add security.syscalls.intercept.mount` commit

File: lxd/include/lxd_seccomp.h [hunk 3]
```diff
@@ -22,6 +22,9 @@
 
 #ifndef SECCOMP_USER_NOTIF_FLAG_CONTINUE
 #define SECCOMP_USER_NOTIF_FLAG_CONTINUE 0x00000001
+#else
+#undef SECCOMP_USER_NOTIF_FLAG_CONTINUE
+#define SECCOMP_USER_NOTIF_FLAG_CONTINUE 0x00000001
```
Comment (stgraber) [inline lxd/include/lxd_seccomp.h]: Seems like this should have belonged to the previous PR, can you make it a separate commit?

File: scripts/bash/lxd-client [hunk 4]
```diff
@@ -94,7 +94,10 @@ _have lxc && {
       security.nesting security.privileged security.protection.delete \
       security.protection.shift security.syscalls.blacklist \
       security.syscalls.blacklist_compat security.syscalls.blacklist_default \
-      security.syscalls.intercept.mknod security.syscalls.intercept.setxattr \
+      security.syscalls.intercept.mknod security.syscalls.intercept.mount \
+      security.syscalls.intercept.mount.allowed \
+      security.syscalls.intercept.setxattr \
+      security.syscall.intercept.mount.shift \
```
Comment (stgraber) [inline scripts/bash/lxd-client]: This should be in a separate `scripts: Add security.syscalls.intercept.mount to bash completion` commit

File: shared/version/api.go [hunk 5]
```diff
@@ -160,6 +160,7 @@ var APIExtensions = []string{
 	"resources_v2",
 	"container_exec_user_group_cwd",
 	"container_syscall_intercept",
+	"container_syscall_intercept_mount",
```
Comment (stgraber) [inline shared/version/api.go]: We only append to this list, so this needs to go at the end and be part of the `api: ...` commit mentioned earlier.

**Comment by brauner** [comment]
Btw, Github is showing the order of the commits wrong. They are in correct order in my branch though, i.e. the implementation first and then the docs.

File: shared/version/api.go [hunk 6]
```diff
@@ -172,6 +172,7 @@ var APIExtensions = []string{
 	"resources_network_firmware",
 	"backup_compression_algorithm",
 	"ceph_data_pool_name",
+	"container_syscall_intercept_mount",
```
Comment (stgraber) [inline shared/version/api.go]: This is in the wrong commit, must be in the API extension commit.
